garrysmod/lua/includes/extensions/net.lua

```diff
@@ -173,11 +139,538 @@ net.ReadVars =
+
+			local valuetype = net.ReadUInt(TYPE_SIZE)
+			local valuevalue = reading[ TypeToString( valuetype ) ]( references )
+
+			reference(valuetype, valuevalue)
+
+			ret[ keyvalue ] = valuevalue
+
+		end
+
+		return ret
+	end
+}
+
+--
+-- We need this since #table returns undefined values
+-- by the lua spec if it doesn't have incremental keys
+-- we use pairs since it's backwards compatible
+--
+local function array_len(x)
+	local indices = {}
+	for k,v in pairs(x) do
+		indices[k] = true
+	end
+
+	for i = 1, 8096 do
+		if(nil == indices[i]) then
+			return i - 1
+		end
+	end
+
+	return 8096
+end
+
+writing = {
+
+	bit      = net.WriteBit,
+	Color    = net.WriteColor,
+	boolean  = net.WriteBool,
+	number   = net.WriteDouble,
+	Entity   = net.WriteEntity,
+	VMatrix  = net.WriteMatrix,
+	Vector   = net.WriteVector,
+	Angle    = net.WriteAngle,
+
+	int16 = function( w ) net.WriteInt( w, 16 ) end,
+	int32 = function( d ) net.WriteInt( d, 32 ) end,
+
+	--
+	-- Variable length unsigned integers
+	--
+	uintv = function( n )
+		while( n > 0 ) do
+			net.WriteBool(true)
+			net.WriteUInt( n, UINTV_SIZE )
+			n = bit.rshift( n, UINTV_SIZE )
+		end
+
+		net.WriteBool(false)
+	end,
+
+
+	--
+	-- 7 bit encoded strings
+	-- NULL terminated
+	--
+	string7 = function( s )
+		local null = s:find( "%z" )
+
+		if (null) then
+			s = s:sub( 1, null - 1 )
+		end
+
+		local compressed = util.Compress( s )
+
+		-- add one for the null terminator
+		if ( compressed and compressed:len() < (s:len() + 1) / 8 * 7 ) then
+			net.WriteBool( true )
+			writing.uintv( compressed:len() )
+			net.WriteData( compressed, compressed:len() )
+		else
+			net.WriteBool( false )
+
+			for i = 1, s:len() do
+				net.WriteUInt( s:byte( i, i ), 7 )
+			end
+
+			net.WriteUInt( 0, 7 )
+		end
+	end,
+
+	--
+	-- Our 6-bit encoded strings
+	-- NULL terminated
+	--
+	hexastring = function( s )
+		local null = s:find( "%z" )
+
+		if (null) then
+			s = s:sub( 1, null - 1 )
+		end
+
+		local compressed = util.Compress( s )
+
+		-- add one for the null terminator
+		if ( compressed and compressed:len() < ( s:len() + 1 ) / 8 * 6 ) then
+			net.WriteBool( true )
+			writing.uintv( compressed:len() )
+			net.WriteData( compressed, compressed:len() )
+		else
+			net.WriteBool( false )
+
+			for i = 1, s:len() do
+				net.WriteUInt( Char2HexaLookup[ s:byte( i, i ) ], 6 )
+			end
+
+			net.WriteUInt( 0, 6 )
+		end
+	end,
+
+	--
+	-- C String
+	-- NULL terminated
+	-- NOTE: Must be NULL terminated or else will break compatibility
+	-- with some addons! Also could lead to exploits
+	--
+	string = function( x )
+		local null = x:find( "%z" )
+
+		if (null) then
+			x = x:sub( 1, null - 1 )
+		end
+
+		local compressed = util.Compress( x )
+
+		if ( compressed and compressed:len() < x:len() + 1 ) then
+			net.WriteBool( true )
+			writing.uintv( compressed:len() )
+			net.WriteData( compressed, compressed:len() )
+		else
+			net.WriteBool( false )
+			net.WriteString( x:len() )
+		end
```

Comment:
When sending, if a string cannot be compressed down with `util.Compress`, then it will write the length of the string instead of the uncompressed string. This can happen with strings that contain multi-byte characters. 

For example: `á` will be received as `2`

```suggestion
		if ( compressed and compressed:len() < x:len() + 1 ) then
			net.WriteBool( true )
			writing.uintv( compressed:len() )
			net.WriteData( compressed, compressed:len() )
		else
			net.WriteBool( false )
			net.WriteString( x )
		end
```

Reply:
This doesn't sound right

Reply:
Try it yourself by networking this table:
```
local strtbl = {
    "aaaaaaaaaaaaaaaaaaaaaaaaaaaaa!",
    "abcdefghijklmnopqrstuvwxyzaaa!",
    "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaz",
    "abcdefghijklmnopqrstuvwxyzaaaa",
    "aaaaaaaaaaaaaaaaaaaaaaaaaaaaa∑",
    "abcdefghijklmnopqrstuvwxyaaaa∑",
}
```

Currently:
![image](https://user-images.githubusercontent.com/22532950/105615785-d97a1180-5d98-11eb-90f6-9bda66f68bd6.png)

After the suggested change:
![image](https://user-images.githubusercontent.com/22532950/105616006-42618980-5d99-11eb-91a3-aa02581506bb.png)

Reply:
Oh, I thought you meant util.Compress was returning the length if it failed to compress. I see the bug in his source now.

Reply:
Ah my bad there. I wasn't very clear.